### ISSUE TITLE
MSI FLASHBIOS feature is not working

### ISSUE THREAD

**Issue by miczyg1** [Dasharo/dasharo-issues#131]
**Dasharo version**
Any

**Dasharo variant**
Dasharo compatible with MSI PRO Z690-A DDR4

**Affected component(s) or functionality**
FLASHBIOS over USB.

**Brief summary**
The FLASHBIOS USB port is not working when using Dasharo

**How reproducible**
Always.

**How to reproduce**

1. Flash Dasharo firmware on the board
2. Format USB stick to FAT32 and put genuine MSI firmware as MSI.ROM
3. Put the USB stick into FLASHBIOS port and press the FLASHBIOS button
4. Observe the process fails by watching the diode.

**Expected behavior**
FLASHBIOS works and updates the firmware.

**Actual behavior**
FLASHBIOS doesn't work

**Screenshots**
None

**Additional context**
None

**Solutions you've tried**
None

**Comment by zirblazer** [comment]
This is a critical feature, as currently, the only two options to flash Dasharo or go back to MSI original Firmware involves either a functional system so that you can use Software flashrom, or having external tools available at hand with the Motherboard exposed and accessible. Getting control of BIOS Flashback will add a really needed intermediate option that will help with bricks caused by bad flashing or incompatible Hardware. Many tears will be shed by users that brick their brand new systems and figure out later on that recovery is hard, and this scares away too many potential users.


Last time this topic was touched, it was discovered than there is a special chip that handles this feature, a Fintek F75504A. It was speculated than it was similar to other chips of its family like the F75501, of which there is a bit of info available: 
https://www.fintek.com.tw/index.php/products/i-o-controller-and-i-f-management/item/133-f75501

Is likely than the F75504A is a Microcontroller with an 8051 CPU plus internal Flash memory that has been factory programmed with an application of sorts. It is wired as a middleman between a specific USB Port and the SPI Bus with access to the Motherboard Firmware Flash EEPROM, and perhaps an extra I2C/SMBus to be visible on the host computer (Don't recall if this was confirmed).
So far, the expected mode of operation is than when the BIOS Flashback button is pushed, it checks both the Motherboard EEPROM and the ROM file on the USB Flash Drive to see if they're matched at some level (Perhaps some string with vendor and model). This is theorized because when the Motherboard is flashed with Dasharo this function stops working, so the Microcontroller most likely does a safety check of some Firmware strings before flashing the ROM file, and it is not known whenever you can force it to flash an arbitrary file.

The plan was to get USB and SPI sniffers to see what was going on when the feature is requested, since knowing what safety values it is checking would allow to test if a custom Dasharo binary that matches the Microcontroller expectations gets flashed. Is very likely than due to logistics, all the F75504A are programmed by MSI the same because it would be easier to deploy them if they're universal instead of having to customize the internal application per Motherboard model, and it will also explain why you can't go from Dasharo to MSI using it. Dumping and reverse engineering the internal Microcontroller Flash could also be viable, but harder, since it means than there has to be some Software way to dump it, but there is no datasheet available to know how the chip works.

**Comment by pietrushnic** [comment]
> This is a critical feature, as currently, the only two options to flash Dasharo or go back to MSI original Firmware involves either a functional system so that you can use Software flashrom, or having external tools available at hand with the Motherboard exposed and accessible. Getting control of BIOS Flashback will add a really needed intermediate option that will help with bricks caused by bad flashing or incompatible Hardware. Many tears will be shed by users that brick their brand new systems and figure out later on that recovery is hard, and this scares away too many potential users.

I agree this makes recovery procedure so cheap that you can freely play with your BIOS what is needed to grow community.

**Comment by Firminator** [comment]
> This is a critical feature, as currently, the only two options to flash Dasharo or go back to MSI original Firmware involves either a functional system so that you can use Software flashrom, or having external tools available at hand with the Motherboard exposed and accessible. Getting control of BIOS Flashback will add a really needed intermediate option that will help with bricks caused by bad flashing or incompatible Hardware. Many tears will be shed by users that brick their brand new systems and figure out later on that recovery is hard, and this scares away too many potential users.
> 
Yes exactly. I ran into this... see https://github.com/Dasharo/dasharo-issues/issues/274 and having Flashback available to go temporarily back to the MSI BIOS would have nice. 

+100 for growth of the platform.

**Comment by Firminator** [comment]
Maybe contacting MSI directly instead of reverse-engineering is a way. Has MSI denied to open-source this? Maybe get media coverage on Phoronix and Reddit to make them reconsider their decision.

**Comment by pietrushnic** [comment]
@Firminator we tried to contact MSI Poland but no matter we asked for the answer is that MSI creates BIOS by itself and is not interested in other BIOS solutions.

IMHO everything is clear, unless we commercial success with Dasharo, we will not get any support.

**Comment by pietrushnic** [comment]
:thinking: why we have it in v1.1.3? I thought we will include it in v1.1.2?

**Comment by miczyg1** [comment]
Yes, the milestones are desynchronized. Plans were different when I have created these milestones. Anyway, fix released in v1.1.2 so closing the issue

**Comment by n3tium** [comment]
> **Dasharo version** Any
> 
> **Dasharo variant** Dasharo compatible with MSI PRO Z690-A DDR4
> 
> **Affected component(s) or functionality** FLASHBIOS over USB.
> 
> **Brief summary** The FLASHBIOS USB port is not working when using Dasharo
> 
> **How reproducible** Always.
> 
> **How to reproduce**
> 
> Steps to reproduce the behavior:
> 
> 1. Flash Dasharo firmware on the board
> 2. Format USB stick to FAT32 and put genuine MSI firmware as MSI.ROM
> 3. Put the USB stick into FLASHBIOS port and press the FLASHBIOS button
> 4. Observe the process fails by watching the diode.
> 
> **Expected behavior** FLASHBIOS works and updates the firmware.
> 
> **Actual behavior** FLASHBIOS doesn't work
> 
> **Screenshots** None
> 
> **Additional context** None
> 
> **Solutions you've tried** None


This chip does not flash the BIOS if it is transplanted to another platform. I installed this chip from b450 to z690 and it flashed the BIOS from b450. I insert the BIOS from the z690 and nothing happens. Those. I think the chip has a built-in check for platform compliance. He most likely does not ask his motherboard for ID.

**Comment by zirblazer** [comment]
> This chip does not flash the BIOS file if it is transplanted to another platform. I installed this chip from b450 to z690 and it flashed the BIOS file from b450. I insert the BIOS file from the z690 and nothing happens. Those. I think the chip has a built-in check for platform compliance. He most likely does not ask his motherboard for ID.

I don't understand. What chip are you talking about? You mean that you desoldered a Flash EEPROM from another board and resoldered it on the Z690-A? What were you trying to fix that forced you to do that on the first place?
The FlashBIOS microcontroller seems to compare certain strings from the current EEPROM with the ROM on the USB drive. This makes the microcontroller "generic" in nature since otherwise it would need to be customized per board. If you put an EEPROM that came from another board into the Z690-A and didn't preflashed it with the proper BIOS, it makes sense that FlashBIOS doesn't work because the strings will never match.

**Comment by joshgarde** [comment]
I found this thread after Googling about MSI's Flashback feature after working on some unrelated BIOS modding for the Gigabyte MPG X670E Carbon. It seems to be the most documented thread I could find about the feature so I'll drop what I found with a logic analyzer wired to the SPI flash -

It appears that before the Flashback microcontroller starts the erase process, it begins by reading at these offsets:
0x01FFFE00 - 0x1FFE50A
0x01B90000 - ? (couldn't be bothered to count, all FF's)
